### ISSUE TITLE
Repeated trail parameters being used with `DaskStorage`

### ISSUE THREAD

**Issue by jrbourbeau** [optuna/optuna#4604]
### Expected behavior

I noticed that when using the `DaskStorage` integration, I always get a smaller number of unique trial parameters than when I run the same optimization not using a `DaskStorage` instance. For example, here's a snippet that runs two identical HPO searches, the first one using `DaskStorage` and the second one using the normal in-memory storage, and then prints out the number of unique trial parameters:

```python
import optuna

from dask.distributed import Client
from dask.distributed import wait
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

optuna.logging.set_verbosity(optuna.logging.WARNING)

def objective(trial):
    X, y = load_digits(return_X_y=True)
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)

    max_depth = trial.suggest_int("max_depth", 2, 10)
    n_estimators = trial.suggest_int("n_estimators", 1, 100)

    clf = RandomForestClassifier(max_depth=max_depth, n_estimators=n_estimators)
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    return score


if __name__ == "__main__":

    with Client() as client:
        print(f"Dask dashboard is available at {client.dashboard_link}")
        storage = optuna.integration.dask.DaskStorage()
        study = optuna.create_study(storage=storage, direction="maximize", sampler=optuna.samplers.RandomSampler(seed=2))
        futures = [
            client.submit(study.optimize, objective, n_trials=1, pure=False) for _ in range(50)
        ]
        wait(futures)
        assert len(study.trials) == 50
        # Get number of unique parameter combinations
        x = study.trials_dataframe()
        x = x[[i for i in x.columns if i.startswith("params_")]]
        num_unique = len(x.drop_duplicates())
        print(f"(dask) {num_unique = }")

    # Now do the same thing, but without using the Dask integration
    study = optuna.create_study(direction="maximize", sampler=optuna.samplers.RandomSampler(seed=2))
    study.optimize(objective, n_trials=50)

    assert len(study.trials) == 50
    # Get number of unique parameter combinations
    x = study.trials_dataframe()
    x = x[[i for i in x.columns if i.startswith("params_")]]
    num_unique = len(x.drop_duplicates())
    print(f"(non-dask) {num_unique = }")
```

Running this locally a few times I get output like

```
(dask) num_unique = 13
(non-dask) num_unique = 49
```

```
(dask) num_unique = 20
(non-dask) num_unique = 49
```

```
(dask) num_unique = 22
(non-dask) num_unique = 49
```

Note that the non-`DaskStorage` version always gives the same number of unique trial parameters (equal to the overall number of trials), while the `DaskStorage` version gives me a much smaller number (it's also not deterministic, even though this example sets `seed=` in the sampler). Naively I would expect these two approaches to give similar results. 

### Environment

- Optuna version: 3.2.0.dev
- Python version: 3.9.9
- OS: macOS-13.3.1-x86_64-i386-64bit

### Error messages, stack traces, or logs

```shell
See "Expected behavior" section above
```

### Steps to reproduce

See the code snippet in the "Expected behavior" section above
